### PR TITLE
Be explicit about constant time bignum functions that must take a 0 or 1 condition value

### PR DESCRIPTION
**Pull request by tom-cosgrove-arm**
## Status
**READY**

## Requires Backporting
Yes - #6212

## Todos
- [X] Documentation
- [X] Backported

### CODE REVIEW

**Comment by tom-cosgrove-arm** [comment]
I wrote test cases, which all pass, so you can see what happens if `cond` != `0` or `1` (in these cases, I used `2`)

```
diff --git a/tests/suites/test_suite_mpi.data b/tests/suites/test_suite_mpi.data
index f29dcabd4..2a063c2ab 100644
--- a/tests/suites/test_suite_mpi.data
+++ b/tests/suites/test_suite_mpi.data
@@ -1893,6 +1893,39 @@ mpi_random_fail:2:"01":MBEDTLS_ERR_MPI_BAD_INPUT_DATA
 MPI random bad arguments: min > N = 1, 0 limb in upper bound
 mpi_random_fail:2:"000000000000000001":MBEDTLS_ERR_MPI_BAD_INPUT_DATA
 
+mbedtls_mpi_safe_cond_assign #1
+mbedtls_mpi_safe_cond_assign:"123456":"654321":0:"123456"
+
+mbedtls_mpi_safe_cond_assign #2
+mbedtls_mpi_safe_cond_assign:"123456":"654321":1:"654321"
+
+mbedtls_mpi_safe_cond_assign #3
+mbedtls_mpi_safe_cond_assign:"-123456":"654321":0:"-123456"
+
+mbedtls_mpi_safe_cond_assign #4
+mbedtls_mpi_safe_cond_assign:"-123456":"654321":1:"654321"
+
+mbedtls_mpi_safe_cond_assign #5 - bad output when cond != 0 or 1
+mbedtls_mpi_safe_cond_assign:"-123456":"654321":2:"-654320"
+
+mbedtls_mpi_safe_cond_assign #6 - bad output when cond != 0 or 1
+mbedtls_mpi_safe_cond_assign:"123456":"-654321":2:"654320"
+
+mbedtls_mpi_safe_cond_swap #1
+mbedtls_mpi_safe_cond_swap:"123456":"654321":0:"123456":"654321"
+
+mbedtls_mpi_safe_cond_swap #2
+mbedtls_mpi_safe_cond_swap:"123456":"654321":1:"654321":"123456"
+
+mbedtls_mpi_safe_cond_swap #3
+mbedtls_mpi_safe_cond_swap:"123456":"-654321":0:"123456":"-654321"
+
+mbedtls_mpi_safe_cond_swap #4
+mbedtls_mpi_safe_cond_swap:"-123456":"654321":1:"654321":"-123456"
+
+mbedtls_mpi_safe_cond_swap #5 - bad output when cond != 0 or 1
+mbedtls_mpi_safe_cond_swap:"-123456":"654321":2:"-654321":"123456"
+
 MPI Selftest
 depends_on:MBEDTLS_SELF_TEST
 mpi_selftest:
diff --git a/tests/suites/test_suite_mpi.function b/tests/suites/test_suite_mpi.function
index 2694a44b6..aef079839 100644
--- a/tests/suites/test_suite_mpi.function
+++ b/tests/suites/test_suite_mpi.function
@@ -1412,6 +1412,61 @@ exit:
 }
 /* END_CASE */
 
+/* BEGIN_CASE */
+void mbedtls_mpi_safe_cond_assign( char *input_l, char *input_r,
+                                   int cond, char *input_X )
+{
+    mbedtls_mpi l, r, X;
+
+    mbedtls_mpi_init( &l );
+    mbedtls_mpi_init( &r );
+    mbedtls_mpi_init( &X );     /* expected result */
+
+    TEST_EQUAL( mbedtls_test_read_mpi( &l, input_l ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &r, input_r ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &X, input_X ), 0 );
+
+    TEST_EQUAL( mbedtls_mpi_safe_cond_assign( &l, &r, cond ), 0 );
+
+    TEST_EQUAL( mbedtls_mpi_cmp_mpi( &l, &X ), 0 );
+
+exit:
+    mbedtls_mpi_free( &l );
+    mbedtls_mpi_free( &r );
+    mbedtls_mpi_free( &X );
+}
+/* END_CASE */
+
+/* BEGIN_CASE */
+void mbedtls_mpi_safe_cond_swap( char *input_l, char *input_r,
+                                 int cond,
+                                 char *input_xl, char *input_xr )
+{
+    mbedtls_mpi l, r, xl, xr;
+
+    mbedtls_mpi_init( &l );
+    mbedtls_mpi_init( &r );
+    mbedtls_mpi_init( &xl );    /* expected l after mbedtls_mpi_safe_cond_swap() */
+    mbedtls_mpi_init( &xr );    /* expected r after mbedtls_mpi_safe_cond_swap() */
+
+    TEST_EQUAL( mbedtls_test_read_mpi( &l, input_l ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &r, input_r ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &xl, input_xl ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &xr, input_xr ), 0 );
+
+    TEST_EQUAL( mbedtls_mpi_safe_cond_swap( &l, &r, cond ), 0 );
+
+    TEST_EQUAL( mbedtls_mpi_cmp_mpi( &l, &xl ), 0 );
+    TEST_EQUAL( mbedtls_mpi_cmp_mpi( &r, &xr ), 0 );
+
+exit:
+    mbedtls_mpi_free( &l );
+    mbedtls_mpi_free( &r );
+    mbedtls_mpi_free( &xl );
+    mbedtls_mpi_free( &xr );
+}
+/* END_CASE */
+
 /* BEGIN_CASE depends_on:MBEDTLS_SELF_TEST */
 void mpi_selftest(  )
 {

```

File: include/mbedtls/bignum.h [hunk 1]
```diff
@@ -300,24 +304,28 @@ int mbedtls_mpi_safe_cond_assign( mbedtls_mpi *X, const mbedtls_mpi *Y, unsigned
...
- *                      if( assign ) mbedtls_mpi_swap( X, Y );
+ *                      if( swap ) mbedtls_mpi_swap( X, Y );
  *                 except that it avoids leaking any information about whether
- *                 the assignment was done or not (the above code may leak
+ *                 the swap was done or not (the above code may leak
  *                 information through branch prediction and/or memory access
  *                 patterns analysis).
  *
+ * \warning        If \p swap is neither 0 nor 1, the result of this function
+ *                 is indeterminate, and both \p X and \p Y might end up with
```
Comment (yanesca) [inline include/mbedtls/bignum.h:320]: How about
```suggestion
 *                 is undefined, and both \p X and \p Y might end up with
```
it is shorter and I think it conveys what we mean here as well.

**Review by yanesca — CHANGES_REQUESTED** [review]
Have a minor suggestion, but other than that it looks good to me.

File: include/mbedtls/bignum.h (hunk 1 above)
Comment (yanesca) [reply include/mbedtls/bignum.h:320]: Well, I think that it is not entirely unlike C undefined. We absolutely give no guarantees whatsoever what will happen if the values are not one of these two. We don’t mean undefined behaviour in the compiler sense, but if the reader makes that mistake that still sends the message we want.

Also, this is a value being undefined, technically that is not the same as undefined behaviour, so it is not a complete overlap in terminology.

But then again, I don't mind leaving it as it is.

**Review by yanesca — APPROVED** [review]
LGTM.

File: include/mbedtls/bignum.h (hunk 1 above)
Comment (gilles-peskine-arm) [reply include/mbedtls/bignum.h:320]: In the C standard, the term for code that produces a valid result but there's no guarantee what this result is is _unspecified_ value. An _indeterminate_ value can be a trap representation, so the code may either produce an unspecified result or behave in some documented way in reaction to a trap. The adjective _undefined_ is reserved for behavior, where a program can do anything including launch demons out of your nose.

So here the C term would be _unspecified_. _Indeterminate_ isn't too far off. _Undefined_ would be misleading.

Comment (tom-cosgrove-arm) [reply include/mbedtls/bignum.h:320]: I will use *unspecified* as the precisely correct term in the future; we can update these as and when we are "in the area"

**Pull request merged**